### DISCUSSION TITLE
Help with computer engineering

### DISCUSSION THREAD

**Discussion by ghost** [community/community#67135] [GitHub Education]
### Hi everyone,

I eventually have a test on Binary and Hexadecimal but I do not remember how to do it.
I also have an IQ < 70

Also, are there any resources to learn logic gates, Binary, Hexadecimal etc

### A suggestion...(optional)

Computer Engineering and programming

**Comment by little-software-engineer** [comment]
Hello,

Just like you, I also had exams on topics such as logic gates and binary systems during my studies, so I'll try to help you in the best way I can. What helped me a lot while learning about conversions from one system to another (e.g., from binary to decimal) was creating tables with the corresponding data for each system I had to work with. I also created tables for logic gates. Here's a quick breakdown of number system conversions for you before we proceed:

**Binary to Decimal Conversion**

Binary is a base-2 number system, meaning it uses only two digits: 0 and 1. To convert a binary number to a decimal number (base-10), follow these steps:

Start from the right: Begin with the rightmost digit of the binary number (the least significant bit).

Assign powers of 2: Assign powers of 2 to each digit, starting with 2^0 for the rightmost digit, then 2^1 for the next digit to the left, 2^2 for the next, and so on, increasing the exponent by 1 for each digit to the left.

Multiply and sum: Multiply each digit by its corresponding power of 2 and sum up the results for all digits.

Calculate the decimal value: The sum from step 3 is the decimal equivalent of the binary number.

Example:
Let's convert the binary number 1001 to decimal.

(1 * 2^3) + (0 * 2^2) + (0 * 2^1) + (1 * 2^0) = 8 + 0 + 0 + 1 = 9

**Decimal to Binary Conversion**

Decimal is a base-10 number system, which means it uses digits 0 to 9. To convert a decimal number to binary, you can use a method called "repeated division by 2." Here are the steps:

Start with the decimal number: Begin with the decimal number you want to convert to binary.

Divide by 2: Divide the decimal number by 2 and record the remainder. The remainder will be either 0 or 1.

Repeat: Continue dividing the quotient obtained in the previous step by 2 and recording the remainders until the quotient becomes 0.

Read the binary result: Read the remainders from bottom to top. This will give you the binary equivalent of the decimal number.

Example:
Let's convert the decimal number 13 to binary.

Step 1: 13 / 2 = 6 remainder 1
Step 2: 6 / 2 = 3 remainder 0
Step 3: 3 / 2 = 1 remainder 1
Step 4: 1 / 2 = 0 remainder 1

Reading the remainders from bottom to top gives us 1101.

**Hexadecimal to Decimal Conversion**

Hexadecimal is a base-16 number system that uses digits 0-9 and the letters A-F (or a-f) to represent values 0-15. To convert a hexadecimal number to a decimal number (base-10), follow these steps:

Start from the right: Begin with the rightmost digit of the hexadecimal number (the least significant digit).

Assign powers of 16: Assign powers of 16 to each digit, starting with 16^0 for the rightmost digit, then 16^1 for the next digit to the left, 16^2 for the next, and so on, increasing the exponent by 1 for each digit to the left.

Convert letters to decimal: If the hexadecimal digit is a letter (A, B, C, D, E, F), replace it with its corresponding decimal value (A=10, B=11, C=12, D=13, E=14, F=15).

Multiply and sum: Multiply each digit (including converted letters) by its corresponding power of 16 and sum up the results for all digits.

Calculate the decimal value: The sum from step 4 is the decimal equivalent of the hexadecimal number.

Example:
Let's convert the hexadecimal number 1A3 to decimal.

(3 * 16^0) + (A * 16^1) + (1 * 16^2) = 3 + (10 * 16) + (1 * 256) = 3 + 160 + 256 = 419

**Decimal to Hexadecimal Conversion**

To convert a decimal number to hexadecimal, follow these steps:

Start with the decimal number: Begin with the decimal number you want to convert to hexadecimal.

Divide by 16: Divide the decimal number by 16 and record both the quotient and the remainder. The remainder will be a digit from 0 to 15.

Repeat: Continue dividing the quotient obtained in the previous step by 16 and recording both the quotient and the remainder until the quotient becomes 0.

Convert remainders to hexadecimal: Replace any remainders greater than 9 with their corresponding hexadecimal letters (10=A, 11=B, 12=C, 13=D, 14=E, 15=F). You can write these hexadecimal letters down just for future reference until you learn them by heart.

Read the hexadecimal result: Read the remainders (including converted letters) from bottom to top. This will give you the hexadecimal equivalent of the decimal number.

Example:
Let's convert the decimal number 419 to hexadecimal.

Step 1: 419 / 16 = 26 remainder 3
Step 2: 26 / 16 = 1 remainder 10 (A)
Step 3: 1 / 16 = 0 remainder 1

Reading the remainders from bottom to top gives us 1A3.

As for logic gates, here's another summary:

**AND Gate (&&)**

The AND gate has two inputs (A and B) and one output (F).
The output is true (1) if and only if both inputs are true (1).
Truth Table:

A | B | F
----------
0 | 0 | 0
0 | 1 | 0
1 | 0 | 0
1 | 1 | 1

**OR Gate (||)**

The OR gate also has two inputs (A and B) and one output (F).
The output is true (1) if at least one of the inputs is true (1).
Truth Table:

A | B | F
----------
0 | 0 | 0
0 | 1 | 1
1 | 0 | 1
1 | 1 | 1

**NOT Gate (!)**

The NOT gate has one input (A) and one output (F).
The output is the just the opposite (complement) of the input.
Truth Table:

A | F
-----
0 | 1
1 | 0

**XOR Gate (^)**

The XOR (exclusive OR) gate has two inputs (A and B) and one output (F).
The output is true (1) if the inputs are different; otherwise, it's false (0).
Truth Table:

A | B | F
----------
0 | 0 | 0
0 | 1 | 1
1 | 0 | 1
1 | 1 | 0

**NAND Gate**

The NAND gate is the opposite of the AND gate.
The output is false (0) only if both inputs are true (1); otherwise, it's true (1).

**NOR Gate**

The NOR gate is the opposite of the OR gate.
The output is true (1) only if both inputs are false (0); otherwise, it's false (0).

**XNOR Gate**

The XNOR (exclusive NOR) gate is the opposite of the XOR gate.
The output is true (1) if both inputs are the same; otherwise, it's false (0).

For further reading, I'd reccommend "Digital Design" by M. Morris Mano and Michael D. Ciletti and "Computer Organization and Design" by David A. Patterson and John L. Hennessy. Also, you can check Reddit and Stack Overflow for more detailed explanations. I wish you the best of luck on your exam!
